Show a UI warning when user does not have permission to update/edit an existing Navigation block

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

Explores part of https://github.com/WordPress/gutenberg/issues/36286#issuecomment-990845661 to display a warning to the user if they have insufficient permissions to _**edit/update**_ the given Navigation block.

This also disallows renaming or deleting the Navigation via the Inspector controls sidebar.

Note this does not cover the creation of brand new Navigations, but rather the scenario whereby another user has created the Navigation and a lower permission user is attempting to edit it.

## ⚠️ String changes post string freeze in WP 5.9 Beta 3

Having spoken with release leads, as requested I have raised [this Trac ticket](https://core.trac.wordpress.org/ticket/54643#ticket). 

## How has this been tested?

- Install a Plugin such as User Switching to afford easy switching between Users.
- Create a Admin and Contributor level user.
- As the Contributor, create a Post.
- As the Admin go to that post and add a Navigation block and add some items and save.
- As the Contributor, go back to the Post. You should see a notice warning that you have insufficient permission to edit.
- As the Admin go to that post and check you can still edit the Navigation as before. You should see no notices.

Also check you cannot delete or rename the Navigation using the Inspector controls under "Advanced".

Note you are not stopped from attempting to update. It might be possible to somewhat disable the component but that will be handled in a follow up.

## Screenshots <!-- if applicable -->

<img width="1510" alt="Screen Shot 2021-12-14 at 15 18 47" src="https://user-images.githubusercontent.com/444434/146026372-1b5084bf-0ac0-4c39-a192-171485bab76c.png">


## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Show warning when user has insufficient permision to edit the given Nav

getdave committed:
Move warning underneath block but allow viewing

getdave committed:
Sync warning display with entity loading

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -152,10 +163,19 @@ function Navigation( {
+				),
 			};
 		},
 		[ clientId ]
```
```suggestion
		[ clientId, ref ]
```

getdave replied:
Fixed in ad57987

adamziel commented:
I like this PR, it preserves everything in the post while limiting just the single restricted functionality. It looks like something that other entity-based blocks could use, too. Perhaps the logic here would make a good `<PermissionsRequired />` component? Multi-entity saving could also leverage the permissions data. What if we declared permissions in the entities config? 🤔 Then each `<EntityProvider` could also render a UI warning.  cc @gziolo @youknowriad

draganescu commented:
IMO we should make the blocks that cannot be edited read only and show the permission message in the inspector or in a flash notification when editing is attempted, so that we don't clutter the layouts so much :) 

Particularly if we'll add this to other entity based blocks.

getdave commented:
Glad you folks believe this is a better route forward.

If we can agree to merge this PR, then I can iterate on disabling the UI iterations. 

In terms of time and scope, can we try and land a solution for the Navigation block only? If that can be abstracted for other blocks then great, but with limited time left I believe we need to focus on the Nav block only.

getdave committed:
Include ref as effect dependency

getdave commented:
@jasmussen @javierarce Pinging you both here to see if you have any opinions on how to show the UI warning to the user.

getdave committed:
Revert unintentional edit to comment placement

getdave committed:
Show permisisons warning using global notices system

getdave committed:
Hide delete and rename inspector items based on perms

getdave commented:
> IMO we should make the blocks that cannot be edited read only and show the permission message in the inspector or in a flash notification when editing is attempted, so that we don't clutter the layouts so much :)

I updated the implementation to use the global `@wordpress/ notices` package.

spacedmonkey commented:
@getdave This does not handle the case where a user edits a post that contains a navigation post that they can not access. 

https://user-images.githubusercontent.com/237508/145848284-c4aec100-786d-452a-ae65-b97705f01f16.mov

In this screencast, you can see an admin add a navigation post then a contributor try and edit that post. The REST API 404s but that is not error handling in the editor.

getdave commented:
> @getdave This does not handle the case where a user edits a post that contains a navigation post that they can not access.

@spacedmonkey I don't understand this. I tested it and I couldn't replicate. Did I miss a step?

https://user-images.githubusercontent.com/444434/145855840-76044f3a-a00f-4ed2-ac0a-f424cc46937e.mp4

- I created a Post as Contributor.
- I switched to Admin and edited the Post, adding a Nav block and saving it.
- I switch back to Contributor.
- I saw the Nav block load correctly but the UI message was displayed.

spacedmonkey commented:
@getdave The difference seems to be that you created a new menu, I used an existing menu that was created as part of FSE. 

I ran my test again, got different buggy results. 
Spinner that never goes 
<img width="1904" alt="Screenshot 2021-12-13 at 19 43 27" src="https://user-images.githubusercontent.com/237508/145877654-2942bd41-2a90-46ab-bd04-13c6c4638f3e.png">

But when I select the block, I do get an error message. 
<img width="1904" alt="Screenshot 2021-12-13 at 19 43 30" src="https://user-images.githubusercontent.com/237508/145877830-a5af3818-9a94-407b-87e2-da4870c69925.png">

getdave commented:
@spacedmonkey Those replication steps didn't work either. I tried:

- As admin, creating Navigation in Site Editor
- switching to a Post which has contributor as author
- adding Nav block and selecting the Nav Menu i created in Site Editor and savings.
- switching to Contributor user
- opening the same Post
- Menu loads as expected and the error shows when the block is selected (intentional)

https://user-images.githubusercontent.com/444434/145969491-9bca560e-4bbd-48cf-8dfe-c1e4c8b1e6ab.mp4


Not sure why you're getting different results here. Could you try rebuilding and then running the test again and noting down the exact steps to replicate what you're experiencing?

One thing I noted was that (for some reason) then Menu you are selecting has no items. Was that intentional?

spacedmonkey commented:
Steps to replicate. 

1. Create a post with navigation as admin. 
2. Login as contributor. 
3. Create post
4. Try to embed navigation for admin post. 
5. See spinner.

spacedmonkey commented:
@getdave 

https://user-images.githubusercontent.com/237508/145990828-35cd0543-5c71-49c4-9f2e-807228bf2919.mov

getdave commented:
@spacedmonkey Thanks. I've now managed to replicate and the bug seems to be specific to your selecting `All Pages` when creating the Menu. That means under the hood we're using the Page List block. 

It looks like it doesn't consider permissions in that many users can't access Pages including contributor users.

I'm not saying having this bug in the experience is _acceptable_, but if you try this PR with _standard_ links it works perfectly.

I have raised [a followup PR to fix the Page List](https://github.com/WordPress/gutenberg/issues/37368) but it's a bug that exists in `trunk` separately to the changes in this PR.

As a result I think we should assess this PR against using standard links and fix the bug with Page List as a seperate issue.

jasmussen commented:
> IMO we should make the blocks that cannot be edited read only and show the permission message in the inspector or in a flash notification when editing is attempted, so that we don't clutter the layouts so much :)

I'm tempted to echo this. The snackbar warning announces well to screen readers, and by virtue of appearing in context, every time you try to make an edit, feels like the better interface for this warning. 

The ideal solution would be for the block to just be read-only, but I recognize the challenges around that.

spacedmonkey reviewed (changes requested):
Before approving this change, it needs some kind of automated test. I think an E2E test seem reasonable.

getdave commented:
> IMO we should make the blocks that cannot be edited read only 

Agreed but it's not feasible in time for the RC.

> ...and show the permission message in the inspector 

I tried that but adding to `<InspectorControls>` within the block's edit doesn't actually place it at top of sidebar UI. This is because global styles items take precedence. Again this will require work _post_-WP 5.9.

> The snackbar warning announces well to screen readers, and by virtue of appearing in context, every time you try to make an edit, feels like the better interface for this warning.

I'm ok with switching to flash notification. I'll get that sorted - glad I checked.

> Before approving this change, it needs some kind of automated test. I think an E2E test seem reasonable.

Agreed. I'll get on to that now.

getdave committed:
Switch to snackbar notice

getdave committed:
Add e2e test covering update permission notice

getdave commented:
Ok so updates have been made:

- @jasmussen I have switched the notice to "Snackbar".
- @spacedmonkey I have added an e2e test. Passes locally. Let's hope it passes on CI. 

What say you? ✅ or 🔴 ?

Follow ups:

- Fix error in Page List block. https://github.com/WordPress/gutenberg/issues/37368
- https://github.com/WordPress/gutenberg/pull/37289

jasmussen commented:
In my quick testing, I think this is good to go as it solves an important problem, and it does not preclude us from making further enhancements in future versions. 👍 👍 from me:
![contributor](https://user-images.githubusercontent.com/1204802/146044364-03ca4b5e-5ae8-495a-846b-6bbd7cd87a98.gif)

_Note that the site logo missing its illustration inside is a separate issue that testing this PR revealed, but is not a blocker since it's unrelated._

Edit: perhaps in nitpick territory, I think we could potentially refine and minimize the snackbar phrase down to "You do not have permissions to edit this block. Edits made will not be saved." — but that's also not a change that should block it. It's fine as is, and we can refine in the future.

spacedmonkey reviewed (approved):
Works well. However, it the snackbar message show everytime the block is selected.

talldan commented:
I think it should be possible to make the blocks mostly non-editable. Possibly it just requires setting `templateLock` to `all`. If that doesn't achieve everything needed it could also be possible to conditionally use the `Disabled` component and set `renderAppender` to `false`.

I think it could be worth trying it out in a separate PR as a small follow-up.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -303,6 +337,53 @@ function Navigation( {
+					'You do not have permission to edit this Navigation. Any edits made will not be saved.'
+				),
+				{
+					id: noticeRef.current,
```
I think a unique id is overkill for this. It just has to be a string that isn't used elsewhere to avoid multiple notifications with the same message. I think you can make it a namespaced string like 'block-library/core/navigation/permissions'.

getdave replied:
OK fixed.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -303,6 +337,53 @@ function Navigation( {
+
+			createWarningNotice(
+				__(
+					'You do not have permission to edit this Navigation. Any edits made will not be saved.'
```
'edit this Navigation' doesn't sound quite right to me, I think that's because in language 'navigation' isn't usually a tangible thing. I also don't think we use this term alone anywhere else.

I think we should qualify this by saying 'edit this Navigation block' or 'edit this Navigation menu', or possibly even just 'edit this block' or 'edit this menu'.

getdave replied:
I think the concept must be editing the Navigation Menu. I'll update that.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -152,13 +168,29 @@ function Navigation( {
 				innerBlocks: blocks,
 				isInnerBlockSelected: hasSelectedInnerBlock( clientId, true ),
+				canUserUpdateNavigationEntity: ref
+					? canUser( 'update', 'navigation', ref )
```
It was mentioned elsewhere, but 'update' is a `PUT` request, which is never actually something we do. I think this should be 'create' which represents the `POST` method.

getdave replied:
So the permissions handling is pretty ~confusing~ involved. Here's my understanding.

- [`canUser('update')`](https://github.com/WordPress/gutenberg/blob/4cedaf0c4abb0aeac4bfd4289d63e9889efe9733/packages/core-data/src/resolvers.js#L271) checks for a `PUT` included in the `Allow` header in the response from the `/wp/navigation/{ID}` endpoint.
- [the `Allow` header is set](https://github.com/WordPress/wordpress-develop/blob/424afb85afa2e19f30d9836440f63c5237b4102e/src/wp-includes/rest-api.php#L797) based on the `permission_callback` of the endpoint.
- the Navigation endpoint defers to the [`WP_REST_Posts_Controller`](https://github.com/WordPress/wordpress-develop/blob/424afb85afa2e19f30d9836440f63c5237b4102e/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php).
- if the endpoint is called with an `ID` then it will be one [the `EDITABLE` methods](https://github.com/WordPress/wordpress-develop/blob/424afb85afa2e19f30d9836440f63c5237b4102e/src/wp-includes/rest-api/class-wp-rest-server.php#L39)(including `PUT` and `POST`) that are [the handlers](https://github.com/WordPress/wordpress-develop/blob/424afb85afa2e19f30d9836440f63c5237b4102e/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L122-L127). 
- the `permission_callback` is [`update_item_permissions_check`](https://github.com/WordPress/wordpress-develop/blob/424afb85afa2e19f30d9836440f63c5237b4102e/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L766) (which is what is called to determine the `Allow` header.

Therefore I _think_ using `update` here is valid. We're checking whether we can edit this _specific_ Navigation Post not whether we can create Posts.

Also worth noting this is [the pattern also in use by the Site Logo block](https://github.com/WordPress/gutenberg/blob/4cedaf0c4abb0aeac4bfd4289d63e9889efe9733/packages/block-library/src/site-logo/edit.js#L329).

spacedmonkey replied:
What @getdave has written here is correct and this looks good to me. Basically anyone can create a post in WordPress. But being able to publish or edit post is what is needed here. Even a contributor can create a post, this just means they can create a draft / pending post. Knowing that isn't especially useful here.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -152,13 +168,29 @@ function Navigation( {
 				hasSubmenus: firstSubmenu,
 				innerBlocks: blocks,
 				isInnerBlockSelected: hasSelectedInnerBlock( clientId, true ),
+				canUserUpdateNavigationEntity: ref
+					? canUser( 'update', 'navigation', ref )
+					: undefined,
+				hasResolvedCanUserUpdateNavigationEntity: hasFinishedResolution(
+					'canUser',
+					[ 'update', 'navigation', ref ]
+				),
+				canUserDeleteNavigationEntity: ref
+					? canUser( 'delete', 'navigation', ref )
+					: undefined,
+				hasResolvedCanUserDeleteNavigationEntity: hasFinishedResolution(
+					'canUser',
+					[ 'delete', 'navigation', ref ]
+				),
```
It might be nice to move these into the `useNavigationMenu` effect and return only two values like `isNavigationMenuEditable` and `isNavigationMenuDeletable` (do the && logic in the effect).

getdave replied:
I moved into the hook.

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -766,4 +770,58 @@ describe( 'Navigation', () => {
+			// Tidy up after ourselves.
+			await deleteUser( username );
+
+			// Expect a console 403 for request to Navigation Areas.
+			// Todo: removed once Nav Areas are removed from the Gutenberg Plugin.
+			expect( console ).toHaveErrored();
```
I thought this had been fixed already? 

edit: Actually, I think navigation areas are supposed to work in the plugin, so not sure why there would be a 403 or why it isn't encountered in other e2e tests.

getdave replied:
Yes they are supposed to work in the Plugin, 

> ...so not sure why there would be a 403...

As with many cases, this user doesn't have _permissions_ to read them. Requires `edit_theme_options` which is effectively only `Admin` roles.

Notice how the error messages match up

https://github.com/WordPress/gutenberg/blob/4cedaf0c4abb0aeac4bfd4289d63e9889efe9733/lib/class-wp-rest-block-navigation-areas-controller.php#L81-L91
<img width="1275" alt="Screen Shot 2021-12-15 at 10 33 52" src="https://user-images.githubusercontent.com/444434/146170695-794ef8db-522f-4a2d-b398-00e6397f49a6.png">

> ....why it isn't encountered in other e2e tests.

That's because no other tests switch to a lower permission users 😞 

I'll add a comment to the test to explain.

getdave replied:
Added in https://github.com/WordPress/gutenberg/pull/37286/commits/cfc7301e32281ad6bd3c3364fac0fabea5e3dbb1

talldan replied:
Right, yep, that makes sense now, a comment would be very helpful, thanks!

talldan committed:
Try removing forward slash to see if test passes

talldan commented:
I think the flaky test system had an issue with the forward slash in the test description, so I've changed the test description. Hopefully that makes it pass.

getdave commented:
> I think it could be worth trying it out in a separate PR as a small follow-up.

Yes let's tackle this separately.

getdave committed:
Remove usage of uniqueId

Resolves https://github.com/WordPress/gutenberg/pull/37286#discussion_r769211397

getdave committed:
Update error message to use clearer terminology

Addressses https://github.com/WordPress/gutenberg/pull/37286#discussion_r769214084

getdave committed:
Fix test to match code error message string change

getdave committed:
Move permissions selectors to existing hook

getdave committed:
Add explaination of requirement for 403 expect for Nav Areas

getdave commented:
> However, it the snackbar message show everytime the block is selected.

Thanks @spacedmonkey. That's actually intended based on @jasmussen's [comment](https://github.com/WordPress/gutenberg/pull/37286#issuecomment-993539504) 

> @jasmussen: The snackbar warning announces well to screen readers, and by virtue of appearing in context, every time you try to make an edit, feels like the better interface for this warning.

ockham commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -766,4 +770,61 @@ describe( 'Navigation', () => {
+			// Tidy up after ourselves.
+			await deleteUser( username );
+
+			// Expect a console 403 for request to Navigation Areas for lower permisison users.
+			// This is because reading requires the `edit_theme_options` capability
+			// which the Contributor level user does not have.
+			// See: https://github.com/WordPress/gutenberg/blob/4cedaf0c4abb0aeac4bfd4289d63e9889efe9733/lib/class-wp-rest-block-navigation-areas-controller.php#L81-L91.
```
Seems similar to https://github.com/WordPress/gutenberg/issues/33003?

getdave replied:
AFAIK it's entirely expected. In this case the contributor user should not have permission and thus 403 is expected. Running the same test as Admin would see no 403 as they have the requisite capability.

getdave committed:
Attempt to fix flaky test on CI

getdave commented:
@talldan I've adjusted as required and @ockham and I fixed the flaky test issue (turns out you need the create user bit in `before/afterAll` hooks.

Are you happy to merge?

adamziel commented on packages/block-library/src/navigation/use-navigation-menu.js:
```diff
@@ -64,6 +65,20 @@ export default function useNavigationMenu( ref ) {
 				),
 				navigationMenu,
 				navigationMenus,
+				canUserUpdateNavigationEntity: ref
+					? canUser( 'update', 'navigation', ref )
+					: undefined,
+				hasResolvedCanUserUpdateNavigationEntity: hasFinishedResolution(
+					'canUser',
+					[ 'update', 'navigation', ref ]
+				),
```
I don't think we need to output a separate `hasResolved` thing:

```suggestion
				canUserUpdateNavigationEntity: ref
					? canUser( 'update', 'navigation', ref ) 
					&& hasFinishedResolution(
						'canUser',
						[ 'update', 'navigation', ref ]
					)
					: undefined,
```

adamziel replied:
Nevermind, we actually need to check this in the component https://github.com/WordPress/gutenberg/pull/37286/files#diff-884dae78f1676d454c049107e9b05b2403eb042457f77f836bd170d104396b8bR350

getdave replied:
I'm confused. You've suggested and then marked as resolved 😕

adamziel commented on packages/block-library/src/navigation/use-navigation-menu.js:
```diff
@@ -64,6 +65,20 @@ export default function useNavigationMenu( ref ) {
+					'canUser',
+					[ 'update', 'navigation', ref ]
+				),
+				canUserDeleteNavigationEntity: ref
+					? canUser( 'delete', 'navigation', ref )
+					: undefined,
+				hasResolvedCanUserDeleteNavigationEntity: hasFinishedResolution(
+					'canUser',
+					[ 'delete', 'navigation', ref ]
+				),
```
Same as above

```suggestion
				canUserDeleteNavigationEntity: ref
					? canUser( 'delete', 'navigation', ref )
					&& hasFinishedResolution(
						'canUser',
						[ 'delete', 'navigation', ref ]
					),
					: undefined,
```

Linter will likely complain about this code formatting, it might need an extra pass

getdave replied:
I'm confused. You've suggested and then marked as resolved 😕

adamziel commented:
I cannot test due to a [npm outage](https://status.npmjs.org/), but the code looks good to me.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -158,7 +162,7 @@ function Navigation( {
 			};
 		},
-		[ clientId ]
+		[ clientId, ref ]
```
Looks like ref is no longer needed as a dep.
```suggestion
		[ clientId ]
```

talldan committed:
Remove ref as a dependency to useSelect

talldan reviewed (approved):
LGTM, thanks for making those late changes.

getdave commented:
Thanks to everyone who helped with this, particularly @spacedmonkey for his REST API knowledge and @talldan and @adamziel for reviewing 🙇‍♂️

getdave commented:
## ⚠️ String changes post string freeze in WP 5.9 Beta 3

Having spoken with release leads, as requested I have raised [this Trac ticket](https://core.trac.wordpress.org/ticket/54643#ticket) to track the need to update strings after the Beta 3.
